Firestore/core/src/local/local_store.cc

```diff
@@ -541,6 +507,144 @@ LruResults LocalStore::CollectGarbage(LruGarbageCollector* garbage_collector) {
   });
 }
 
+bool LocalStore::HasNewerBundle(const bundle::BundleMetadata& metadata) {
+  return persistence_->Run("Has newer bundle", [&] {
+    absl::optional<bundle::BundleMetadata> cached_metadata =
+        bundle_cache_->GetBundleMetadata(metadata.bundle_id());
+    return cached_metadata.has_value() &&
+           cached_metadata->create_time() >= metadata.create_time();
+  });
+}
+
+void LocalStore::SaveBundle(const bundle::BundleMetadata& metadata) {
+  return persistence_->Run(
+      "Save bundle", [&] { bundle_cache_->SaveBundleMetadata(metadata); });
+}
+
+MaybeDocumentMap LocalStore::ApplyBundledDocuments(
+    const MaybeDocumentMap& bundled_documents, const std::string& bundle_id) {
+  // Allocates a target to hold all document keys from the bundle, such that
+  // they will not get garbage collected right away.
+  TargetData umbrella_target = AllocateTarget(NewUmbrellaTarget(bundle_id));
+  return persistence_->Run("Apply bundle documents", [&] {
+    DocumentKeySet keys;
+    DocumentUpdateMap document_updates;
+    DocumentVersionMap versions;
+
+    for (const auto& kv : bundled_documents) {
+      const DocumentKey& key = kv.first;
+      const auto& doc = kv.second;
+      if (doc.type() == MaybeDocument::Type::Document) {
+        keys = keys.insert(key);
+      }
+      document_updates.emplace(key, doc);
+      versions.emplace(key, doc.version());
+    }
+
+    target_cache_->RemoveMatchingKeysForTarget(umbrella_target.target_id());
+    target_cache_->AddMatchingKeys(keys, umbrella_target.target_id());
+
+    auto changed_docs = PopulateDocumentChanges(document_updates, versions,
```

Comment:
Nit: the invocation of this function above spells out the type explicitly. I don't mind either, but perhaps use a consistent approach?